packages/salesforcedx-vscode-core/src/commands/forceApexLogFetch.ts

```diff
@@ -0,0 +1,182 @@
+  ): Promise<void> {
+    const cancellationTokenSource = new vscode.CancellationTokenSource();
+    const cancellationToken = cancellationTokenSource.token;
+    const execution = new CliCommandExecutor(this.build(response.data), {
+      cwd: vscode.workspace.rootPath
+    }).execute(cancellationToken);
+    this.attachExecution(execution, cancellationTokenSource, cancellationToken);
+    const resultPromise = new CommandOutput().getCmdResult(execution);
+    const result = await resultPromise;
+    const resultJson = JSON.parse(result);
+    if (resultJson.status === 0) {
+      const logDir = path.join(
+        vscode.workspace.workspaceFolders![0].uri.fsPath,
+        '.sfdx',
+        'tools',
+        'debug',
+        'logs'
+      );
+      if (!fs.existsSync(logDir)) {
+        mkdir('-p', logDir);
+      }
+      const date = new Date(response.data.startTime)
+        .toISOString()
+        .replace('T', ' ')
+        .replace('.000', '');
+      fs.writeFile(
+        path.join(logDir, `${response.data.id}_${date}.log`),
+        resultJson.result.log
+      );
+    }
+  }
+}
+
+export enum ApexDebugLogRequest {
+  Api = 'Api',
+  Application = 'Application'
+}
+
+export type ApexDebugLogIdStartTime = {
+  id: string;
+  startTime: string;
+};
+
+export type ApexDebugLogObject = {
+  Id: string;
+  StartTime: string;
+  LogLength: number;
+  Operation: string;
+  Request: string;
+};
+
+export async function getLogs(
+  cancellationTokenSource: vscode.CancellationTokenSource
+): Promise<ApexDebugLogObject[]> {
+  const execution = new CliCommandExecutor(
+    new SfdxCommandBuilder()
+      .withDescription(nls.localize('force_apex_log_list_text'))
+      .withArg('force:apex:log:list')
+      .withArg('--json')
+      .build(),
+    { cwd: vscode.workspace.workspaceFolders![0].uri.fsPath }
+  ).execute();
+  CancellableStatusBar.show(execution, cancellationTokenSource);
+  taskViewService.addCommandExecution(execution, cancellationTokenSource);
+  const cmdOutput = new CommandOutput();
+  const result = await cmdOutput.getCmdResult(execution);
+  try {
+    const apexDebugLogObjects: ApexDebugLogObject[] = JSON.parse(result).result;
+    apexDebugLogObjects.sort(function(a, b): number {
+      return new Date(b.StartTime).valueOf() - new Date(a.StartTime).valueOf();
+    });
+    return Promise.resolve(apexDebugLogObjects);
+  } catch (e) {
+    return Promise.reject(e);
+  }
+}
+
+export interface ApexDebugLogItem extends vscode.QuickPickItem {
+  id: string;
+}
+
+export class LogFileSelector
+  implements ParametersGatherer<ApexDebugLogIdStartTime> {
+  public async gather(): Promise<
+    CancelResponse | ContinueResponse<ApexDebugLogIdStartTime>
+  > {
+    const cancellationTokenSource = new vscode.CancellationTokenSource();
+    const logInfos = await getLogs(cancellationTokenSource);
+    if (logInfos.length > 0) {
+      const logItems = logInfos.map(logInfo => {
+        let icon = '$(alert) ';
```

Comment:
@ruthemmanuelle @DatGuyJonathan what images do we want to use for RequestTypes? There are two: Application and Api

Reply:
I'm not sure what you're asking. Can you include a screenshot?

Reply:
"VS Code supports icons in quick pick list; I've used it before in exception breakpoints using an icon from https://octicons.github.com/. We could use an icon to describe ApexLog.Request. Currently, there are only two: Api or Application."

![image](https://user-images.githubusercontent.com/4503716/36287885-962ce9cc-126c-11e8-9ba2-82f07268913d.png)

Reply:
Also thinking we should maybe format the time stamp a little differently and indicate that the number at the bottom of each option is the size of the log.

Reply:
Why do we want to show the size of the log? Does that help the user to distinguish one from another? Might be better to not show too much unnecessary information.

Reply:
Can we use the file icon? That would match what we used to represent logging in the status bar. https://octicons.github.com/icon/file

Reply:
Where did you get the name Api and Application from? I only know of them as Monitoring, System, and Preserved, as defined at 

https://developer.salesforce.com/docs/atlas.en-us.api_tooling.meta/api_tooling/tooling_api_objects_apexlog.htm

```
Specifies the location of the origin of the log or heap dump. Values are:
Monitoring — Generated as part of debug log monitoring and visible to all administrators. These types of logs are maintained until the user or the system overwrites them.
SystemLog — Generated as part of system log monitoring and visible only to you. These types of logs are only maintained for 60 minutes or until the user clears them.
Preserved — A system log that is maintained longer than 60 minutes. Used for internal support.
```

Reply:
@vazexqi Size of the log was put as an AC and the reason for it in the work item was to let the user know how much they're about to download.

Api and Application are in the Request section of the doc you linked and were also mentioned in the AC. Monitoring, System, and Preserved are in the Location section.

Reply:
Thanks @Sweetman. Let's revise them a bit based on what the UI looks like and what we did in #303.

1. In your screenshot, could you update to show me something that is an Application and an API request. I am trying to see if the `operation` field is sufficient to distinguish the logs and we don't need to care about different icons for Application vs API since I don't think there are distinct/intuitive icons to represent the two.
1. I am thinking that the time format should match what we have on the hover and the status bar in human format. 
1. The download size should have a unit for it. I think it makes more sense to be in KB rather than Bytes.

Once you do that could you update the screenshots? Thanks!

CC: @DatGuyJonathan

Reply:
@vazexqi Here is a screenshot with a more matching time format, KB size, and API/Application (rocket/squirrel)
![image](https://user-images.githubusercontent.com/4503716/36455974-7f22f386-1657-11e8-9456-f432abfe91f5.png)

Reply:
@Sweetman - I'm not sure if the comment the other day of using the squirrel was serious ;-)

The size of the log in KB looks quite weird. For some reason, it's rendered even bigger than the main information (which is what we can use to filter on as we type).

I suggest the following layout

```
<log icon that matches what we have in status bar> <description - this is what can be used to filter>
<text that says application/api> <date> <time> <size>
```